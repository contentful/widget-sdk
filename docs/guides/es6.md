Writing ES6
===========

Angular services can no can be written using ES6 syntax and are transpiled with
[Babel][babel].

ES6 transpilation is enabled for files with the `.es6.js` extension. See the
[ES6 Modules](#es6-modules) section for more info on how to integrate ES6 code
with Angular.

We use a restricted set of Babel transform plugins that we will extend when
necessary. See `tools/app-babel-options.js` for the list of supported plugins.

ES6 syntax is not yet supported for directives and controllers.

[babel]: http://babeljs.io/


Adding a Transform Plugin
-------------------------

To add support for an ES6 feature one needs to adjust the Babel configuration
and the ESLint configuration. The babel configuration is located in
`tools/app-babel-options.js`

Be sure to prefer the `loose` option for plugins if available. This will create
less correct code but it will be more portable without using polyfills.


Linting
-------

We have to different ESLint configurations `src/javascripts/.eslint-es5.yml` and
`src/javascripts/.eslint-es6.yml` for ES5 and ES6 files, repectively.

The ES6 configuration excludes ES6 syntax that is not transformed through the
[`no-restricted-syntax` rule][no-restricted-syntax-rule]. If you add a Babel
plugin make sure to also update the ESLint configuration.

We use two shell scripts `bin/lint-all` and `bin/lint-file` to lint against the
appropriate configuraiton. You can use the latter to integrate ESLint into your
code editor or IDE.

[no-restricted-syntax-rule]: http://eslint.org/docs/rules/no-restricted-syntax


ES6 Modules
-----------

The ES6 module syntax (using the keywords `import` and `export`) is transpiled
into code that is then plugged into Angularâ€™s DI module system.

Assume the file `src/javascripts/a/b/x.es6.js` has the following export
~~~js
export default const def = 'default'
export const foo = 'foo'
~~~
This file is transformed into an Angular service with its name `a/b/x` derived
from the file name. Thus, from Angular we can import this file as follows.
~~~js
var X = $injector.get('a/b/x')
assert.equal(X.default, 'default')
assert.equal(X.foo, 'foo')
~~~

Conversely, we can import Angular services in ES6 modules.
~~~js
import * as X from 'x'
~~~
is equivalent to
~~~
var X = $injector.get('x')
~~~

We also support relative imports. Assume the file `src/javascripts/a/b/x.es6.js`
has the following import.
~~~js
import * as Y from '../y'
import * as Z from './z'
~~~
Imports are resolved relative to the module ID (which coincides with the service
name) of the file they are used in. Since the file above has the module ID
`a/b/c` this is equivalent to
~~~js
var Y = $injector.get('a/y')
var Z = $injector.get('a/b/z')
~~~

The implementation uses the [SystemJS Babel plugin][babel-systemjs] and is
provided in `src/javascripts/prelude.js`.

[babel-systemjs]: http://babeljs.io/docs/plugins/transform-es2015-modules-systemjs/
