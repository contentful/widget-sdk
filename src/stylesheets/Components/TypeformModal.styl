.cf-typeform-modal
  height: 450px
  padding: 0
  margin: -1.875rem
  > div
    height: 100%
    overflow-y: hidden
