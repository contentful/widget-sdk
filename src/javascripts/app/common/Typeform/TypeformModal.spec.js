import React from 'react';
import Enzyme from 'enzyme';
import { Modal } from '@contentful/forma-36-react-components';
import { TypeformEmbed } from './TypeformEmbed.es6';
import { TypeformModal } from './TypeformModal.es6';

describe('TypeformModal', () => {
  const props = {
    title: 'My typeform modal',
    isShown: true,
    onClose: () => 'close modal',
    typeformUrl: 'https://contentful.typeform.com/potato',
    onTypeformSubmit: () => 'typeform submitted',
    testId: 'my-typeform-modal'
  };
  const wrapper = Enzyme.shallow(<TypeformModal {...props} />);

  it('should pass Modal props to Modal component', () => {
    const WrappingModal = wrapper.find(Modal);

    expect(WrappingModal.prop('title')).toBe(props.title);
    expect(WrappingModal.prop('isShown')).toBe(props.isShown);
    expect(WrappingModal.prop('onClose')).toBe(props.onClose);
    expect(WrappingModal.prop('testId')).toBe(props.testId);
  });

  it('should pass TypeformEmbed props to TypeformEmbed component', () => {
    const EmbeddedTypeform = wrapper.find(TypeformEmbed);

    expect(EmbeddedTypeform.prop('url')).toBe(props.typeformUrl);
    expect(EmbeddedTypeform.prop('onSubmit')).toBe(props.onTypeformSubmit);
  });

  it('should render a Modal with TypeformEmbed inside', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
