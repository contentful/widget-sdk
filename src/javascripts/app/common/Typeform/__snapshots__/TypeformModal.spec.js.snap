// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeformModal should render a Modal with TypeformEmbed inside 1`] = `
.emotion-0 {
  height: 450px;
  padding: 0;
  margin: -1.875rem;
}

.emotion-0 > div {
  height: 100%;
  overflow-y: hidden;
}

<Modal
  allowHeightOverflow={true}
  isShown={true}
  onAfterOpen={[Function]}
  onClose={[Function]}
  position="center"
  shouldCloseOnEscapePress={true}
  shouldCloseOnOverlayClick={true}
  size="700px"
  testId="my-typeform-modal"
  title="My typeform modal"
  topOffset="50px"
>
  <TypeformEmbed
    className="emotion-0"
    hideFooter={true}
    hideHeaders={true}
    onSubmit={[Function]}
    popupAutoCloseDuration={5000}
    popupAutoOpen={false}
    popupMode="popup"
    renderAs="widget"
    url="https://contentful.typeform.com/potato"
    widgetOpacity={0}
    widgetStartButtonText="Start"
  />
</Modal>
`;
